lib/mongo/retryable/write_worker.rb

```diff
@@ -0,0 +1,366 @@
+      # is being committed or aborted, the operation is executed exactly once.
+      # Note that, since transactions require sessions, this method will raise
+      # ArgumentError if ending_transaction is true and session is nil.
+      #
+      # @api private
+      #
+      # @example Execute the write.
+      #   write_with_retry do
+      #     ...
+      #   end
+      #
+      # @note This only retries operations on not master failures, since it is
+      #   the only case we can be sure a partial write did not already occur.
+      #
+      # @param [ nil | Hash | WriteConcern::Base ] write_concern The write concern.
+      # @param [ true | false ] ending_transaction True if the write operation is
+      #   abortTransaction or commitTransaction, false otherwise.
+      # @param [ Context ] context The context for the operation.
+      # @param [ Proc ] block The block to execute.
+      #
+      # @yieldparam [ Connection ] connection The connection through which the
+      #   write should be sent.
+      # @yieldparam [ Integer ] txn_num Transaction number (NOT the ACID kind).
+      # @yieldparam [ Operation::Context ] context The operation context.
+      #
+      # @return [ Result ] The result of the operation.
+      #
+      # @since 2.1.0
+      def write_with_retry(write_concern, ending_transaction: false, context:, &block)
+        session = context.session
+
+        ensure_valid_state!(ending_transaction, session)
+
+        unless ending_transaction || retry_write_allowed?(session, write_concern)
+          return legacy_write_with_retry(nil, context: context, &block)
+        end
+
+        # If we are here, session is not nil. A session being nil would have
+        # failed retry_write_allowed? check.
+
+        server = select_server(cluster, ServerSelector.primary, session)
+
+        unless ending_transaction || server.retry_writes?
+          return legacy_write_with_retry(server, context: context, &block)
+        end
+
+        modern_write_with_retry(session, server, context, &block)
+      end
+
+      # Retryable writes wrapper for operations not supporting modern retryable
+      # writes.
+      #
+      # If the driver is configured to use modern retryable writes, this method
+      # yields to the passed block exactly once, thus not retrying any writes.
+      #
+      # If the driver is configured to use legacy retryable writes, this method
+      # delegates to legacy_write_with_retry which performs write retries using
+      # legacy logic.
+      #
+      # @param [ nil | Hash | WriteConcern::Base ] write_concern The write concern.
+      # @param [ Context ] context The context for the operation.
+      #
+      # @yieldparam [ Connection ] connection The connection through which the
+      #   write should be sent.
+      # @yieldparam [ nil ] txn_num nil as transaction number.
+      # @yieldparam [ Operation::Context ] context The operation context.
+      #
+      # @api private
```

Comment:
This tag is unnecessary here since the whole class is marked with `api private`.